### PR TITLE
[REVIEW] Reading multiple ORC stripes from Python

### LINKED ISSUE #6022: [FEA] Selecting multiple ORC stripes to read from Python
The Python API currently only supports selecting a single stripe.
```python
cudf.read_orc("diamonds.orc", stripe=0)
```
With the proposed change, multiple stripes could be selected, in a way to similar to how multiple Parquet row groups can be selected for reading.
```python
cudf.read_orc("diamonds.orc", stripes=[0, 1, 2])
cudf.read_parquet("diamonds.parquet", row_groups=[[0, 1]])
```

Closed by #6024

### PR DESCRIPTION
**Pull request by calebwin**
This PR exposes selecting multiple ORC stripes from the Python API.

Resolves #6022

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

**Comment by GPUtester** [comment]
Can one of the admins verify this patch?

**Comment by kkraus14** [comment]
add to whitelist

File: python/cudf/cudf/io/orc.py [hunk 1]
```diff
@@ -30,6 +30,7 @@ def read_orc(
     columns=None,
     stripe=None,
     stripe_count=None,
+    stripes=None,
```
Comment (kkraus14) [inline python/cudf/cudf/io/orc.py:33]: Can we have just `stripes` and not support `stripe` / `stripe_count` anymore?

Comment (calebwin) [reply python/cudf/cudf/io/orc.py:33]: Well, the C++ API already supported `stripe_list`, `stripe`, and `stripe_count` [1]. This PR is actually just exposing the `stripe_list`. If you still think `stripe`/`stripe_count` should be discontinued, should the change be made on this PR or a separate one? and should it be made only in the Python API or also the C++ API?

[1] https://github.com/rapidsai/cudf/blob/3cf6c25714e6d0b5ede8a80e6440b647b8f87c35/cpp/include/cudf/io/functions.hpp#L355-L360

Comment (kkraus14) [reply python/cudf/cudf/io/orc.py:33]: We should follow what we did on the parquet side, which I believe is remove the `stripe` and `stripe_count` APIs.

Comment (calebwin) [reply python/cudf/cudf/io/orc.py]: Sounds good. So I removed `stripe` and stripe_count` from the Python API [1] and from the C++ API [2]. Let me know if this looks okay.

[1] https://github.com/rapidsai/cudf/pull/6024/commits/806b932bd221a05533455376dd44eeac27b4ed50
[2] https://github.com/rapidsai/cudf/pull/6024/commits/7f32886a3091601462d322479e7b9e78fceb3f55

**Comment by calebwin** [comment]
@kkraus14 Do I need to do something about the failing check-style and/or pr-builder? It seems like check-style failure is not a style issue.

**Comment by rgsl888prabhu** [comment]
> @kkraus14 Do I need to do something about the failing check-style and/or pr-builder? It seems like check-style failure is not a style issue.

It is just a CI failure, just have to rerun the tests.

File: python/cudf/cudf/tests/test_orc.py [hunk 2]
```diff
@@ -202,13 +202,21 @@ def test_orc_read_stripe(datadir, engine):
...
     ]
     gdf = cudf.concat(gdf).reset_index(drop=True)
+    assert_eq(pdf, gdf, check_categorical=False)
 
+    # Read stripes all at once
+    gdf = cudf.read_orc(path, engine=engine, stripes=range(stripes))
     assert_eq(pdf, gdf, check_categorical=False)
 
+    # Read only some stripes
+    gdf = cudf.read_orc(path, engine=engine, stripes=[0, 1])
```
Comment (kkraus14) [inline python/cudf/cudf/tests/test_orc.py:217]: Can we test reading some non-contiguous stripes as well? I.E. `stripes=[1, 3]`?

Comment (calebwin) [reply python/cudf/cudf/tests/test_orc.py:217]: I added a test for this case and updated the Python documentation comment.

File: cpp/src/io/orc/reader_impl.cu [hunk 3]
```diff
@@ -619,7 +607,6 @@ reader::impl::impl(std::unique_ptr<datasource> source,
 
 table_with_metadata reader::impl::read(size_type skip_rows,
                                        size_type num_rows,
-                                       size_type stripe,
                                        size_type max_stripe_count,
                                        const size_type *stripe_indices,
```
[the remark this answers is not in the document]
Comment (vuule) [reply cpp/src/io/orc/reader_impl.cu]: :fire: :fire: commit!

**Comment by vuule** [comment]
@calebwin you need to update dask-cudf to avoid using the removed parameters. Should be similar to the changes I made in [#5536](https://github.com/rapidsai/cudf/pull/5536/files#diff-a688a35a5060a2f2bf94cfe1dc670f91)

**Comment by calebwin** [comment]
Thanks @vuule. I updated dask-cudf accordingly.

**Comment by kkraus14** [comment]
rerun tests

File: cpp/src/io/orc/reader_impl.cu [hunk 4]
```diff
@@ -158,41 +158,26 @@ class metadata {
...
+   * @param[in] stripe_list Indices of individual stripes
    * @param[in] row_start Starting row of the selection
    * @param[in,out] row_count Total number of rows selected
    *
    * @return List of stripe info and total number of selected rows
    **/
-  auto select_stripes(size_type stripe,
-                      size_type max_stripe_count,
-                      const size_type *stripe_indices,
+  auto select_stripes(const std::vector<size_type> &stripe_list,
```
Comment (harrism) [inline cpp/src/io/orc/reader_impl.cu:167]: It's not a list, so suggest not calling it a list.

```suggestion
  auto select_stripes(const std::vector<size_type> &stripes,
```

File: cpp/src/io/orc/reader_impl.cu [hunk 5]
```diff
@@ -619,17 +604,14 @@ reader::impl::impl(std::unique_ptr<datasource> source,
 
 table_with_metadata reader::impl::read(size_type skip_rows,
                                        size_type num_rows,
-                                       size_type stripe,
-                                       size_type max_stripe_count,
-                                       const size_type *stripe_indices,
+                                       const std::vector<size_type> &stripe_list,
```
Comment (harrism) [inline cpp/src/io/orc/reader_impl.cu:607]:
```suggestion
                                       const std::vector<size_type> &stripes,
```

File: cpp/src/io/orc/reader_impl.hpp [hunk 6]
```diff
@@ -69,18 +69,14 @@ class reader::impl {
...
    * @param stream CUDA stream used for device memory operations and kernel launches.
    *
    * @return The set of columns along with metadata
    */
   table_with_metadata read(size_type skip_rows,
                            size_type num_rows,
-                           size_type stripe,
-                           size_type max_stripe_count,
-                           const size_type *stripe_indices,
+                           const std::vector<size_type> &stripe_list,
```
Comment (harrism) [inline cpp/src/io/orc/reader_impl.hpp:79]:
```suggestion
                           const std::vector<size_type> &stripes,
```

File: python/cudf/cudf/_lib/orc.pyx [hunk 7]
```diff
@@ -133,8 +130,7 @@ cdef read_orc_args make_read_orc_args(filepath_or_buffer,
     cdef read_orc_args args = read_orc_args(
         make_source_info([filepath_or_buffer])
     )
-    args.stripe = stripe
-    args.stripe_count = stripe_count
+    args.stripe_list = stripes
```
Comment (harrism) [inline python/cudf/cudf/_lib/orc.pyx:133]:
```suggestion
    args.stripes = stripes
```

**Review by harrism — CHANGES_REQUESTED** [review]
Just suggesting not naming a vector a list. Remember to update the doxygen @params as well.

File: python/cudf/cudf/_lib/orc.pyx (hunk 7 above)
Comment (calebwin) [reply python/cudf/cudf/_lib/orc.pyx]: Thanks, Mark. I updated this in all appropriate places.

**Pull request merged**